src/main/routing/network_graph.rs

```diff
@@ -0,0 +1,809 @@
+            let node_2 = *graph.node_id_to_index(2).unwrap();
+
+            let shortest_paths = graph.compute_shortest_paths(&[node_0, node_1, node_2]);
+
+            let lookup_latency = |a, b| shortest_paths.get(&(a, b)).unwrap().latency_ns;
+
+            if *directed {
+                assert_eq!(lookup_latency(node_0, node_0), 8);
+                assert_eq!(lookup_latency(node_0, node_1), 3);
+                assert_eq!(lookup_latency(node_0, node_2), 7);
+                assert_eq!(lookup_latency(node_1, node_0), 5);
+                assert_eq!(lookup_latency(node_1, node_1), 8);
+                assert_eq!(lookup_latency(node_1, node_2), 12);
+                assert_eq!(lookup_latency(node_2, node_0), 16);
+                assert_eq!(lookup_latency(node_2, node_1), 11);
+                assert_eq!(lookup_latency(node_2, node_2), 23);
+            } else {
+                assert_eq!(lookup_latency(node_0, node_0), 6);
+                assert_eq!(lookup_latency(node_0, node_1), 3);
+                assert_eq!(lookup_latency(node_0, node_2), 7);
+                assert_eq!(lookup_latency(node_1, node_0), 3);
+                assert_eq!(lookup_latency(node_1, node_1), 6);
+                assert_eq!(lookup_latency(node_1, node_2), 10);
+                assert_eq!(lookup_latency(node_2, node_0), 7);
+                assert_eq!(lookup_latency(node_2, node_1), 10);
+                assert_eq!(lookup_latency(node_2, node_2), 14);
+            }
+        }
+    }
+}
+
+mod export {
+    use super::*;
+
+    #[no_mangle]
+    pub extern "C" fn networkgraph_load(
+        config: *const configuration::ConfigOptions,
+    ) -> *mut NetworkGraph {
+        assert!(!config.is_null());
+        let config = unsafe { &*config };
+
+        match load_network_graph(config.network.graph.as_ref().unwrap()) {
+            Ok(graph_str) => Box::into_raw(Box::new(NetworkGraph::parse(&graph_str).unwrap())),
+            Err(err) => {
+                error!("{}", err);
+                std::ptr::null_mut()
+            }
+        }
+    }
+
+    #[no_mangle]
+    pub extern "C" fn networkgraph_free(graph: *mut NetworkGraph) {
+        assert!(!graph.is_null());
+        unsafe { Box::from_raw(graph) };
+    }
+
+    /// Get the downstream bandwidth of the graph node.
+    #[no_mangle]
+    pub extern "C" fn networkgraph_nodeBandwidthDownBits(
+        graph: *mut NetworkGraph,
+        node_id: u32,
+        bandwidth_down: *mut u64,
+    ) -> bool {
+        let graph = unsafe { graph.as_ref() }.unwrap();
+        let bandwidth_down = unsafe { bandwidth_down.as_mut() }.unwrap();
+
+        let node = graph.node_id_to_index(node_id).unwrap();
+
+        match graph.graph().node_weight(*node).unwrap().bandwidth_down {
+            Some(x) => {
+                *bandwidth_down = x.convert(units::SiPrefixUpper::Base).unwrap().value();
+                return true;
+            }
+            None => return false,
```

Comment:
> Can this happen outside of a bug in Shadow?

We decided not to require bandwidth values on the graph node. We only require that the bandwidth values are provided somewhere, either from the graph node or the config host. It's only an error if neither do.

> As before it'd be nice if we could add the warn_unused_result attribute, but IDK if there's a way to do it with bindgen.

Both bindgen and cbindgen have support for this.

> Consider logging a warning or error, and/or assigning bandwidth_down something like 0 in this case to help make failures here a bit more detectable.

Since it's optional, I don't think it makes sense to log a warning. The goal of this function was to return true if it was set and update the `bandwidth_down` pointer, and return false if it wasn't set. These return values weren't meant to indicate errors, but I could see how it could get confused as an error value.

I don't think we should return 0 here since if we make "unset" equal 0, then if the user does enter 0 for the host bandwidth, shadow will assume it was unset and will use the node bandwidth instead. This might be weird behaviour if the user explicitly set a bandwidth of 0 for the host, but the host is actually given a bandwidth greater than 0 since shadow wouldn't be able to distinguish between 0 and "unset".

Would it be better to change the function to return 0/-1 instead? We could instead make shadow exit() immediately within the rust code if a bandwidth of 0 was set so that we don't have conflicting uses of the 0 value.

Reply:
> Both bindgen and cbindgen have support for this.

Oh cool!

> Since it's optional, I don't think it makes sense to log a warning...

Ah ok. I was thinking of this as a function for core shadow code to get the bandwidth value for a host; I missed the context that this is specifically checking to see if it's on the provided network graph and could alternatively come from the host config.

The current behavior makes sense to me now.

Reply:
I made this return an int (0 if it was found, -1 if it wasn't found) instead of the true/false just to be more consistent. Let me know if you think true/false is better.